lib/config/config-loader.js

```diff
@@ -0,0 +1,718 @@
+        }
+
+        const config = await jiti.import(fileURL.href);
+
+        importedConfigFileModificationTime.set(filePath, mtime);
+
+        return config?.default ?? config;
+    }
+
+
+    // fallback to normal runtime behavior
+
+    const config = (await import(fileURL)).default;
+
+    importedConfigFileModificationTime.set(filePath, mtime);
+
+    return config;
+}
+
+/**
+ * Calculates the config array for this run based on inputs.
+ * @param {string} configFilePath The absolute path to the config file to use if not overridden.
+ * @param {string} basePath The base path to use for relative paths in the config file.
+ * @param {ConfigLoaderOptions} options The options to use when loading configuration files.
+ * @returns {Promise<FlatConfigArray>} The config array for `eslint`.
+ */
+async function calculateConfigArray(configFilePath, basePath, options) {
+
+    const {
+        cwd,
+        baseConfig,
+        ignoreEnabled,
+        ignorePatterns,
+        overrideConfig,
+        defaultConfigs = [],
+        allowTS
+    } = options;
+
+    debug(`Calculating config array from config file ${configFilePath} and base path ${basePath}`);
+
+    const configs = new FlatConfigArray(baseConfig || [], { basePath, shouldIgnore: ignoreEnabled });
+
+    // load config file
+    if (configFilePath) {
+
+        debug(`Loading config file ${configFilePath}`);
+        const fileConfig = await loadConfigFile(configFilePath, allowTS);
+
+        if (Array.isArray(fileConfig)) {
+            configs.push(...fileConfig);
+        } else {
+            configs.push(fileConfig);
+        }
+    }
+
+    // add in any configured defaults
+    configs.push(...defaultConfigs);
+
+    // append command line ignore patterns
+    if (ignorePatterns && ignorePatterns.length > 0) {
+
+        let relativeIgnorePatterns;
+
+        /*
+         * If the config file basePath is different than the cwd, then
+         * the ignore patterns won't work correctly. Here, we adjust the
+         * ignore pattern to include the correct relative path. Patterns
+         * passed as `ignorePatterns` are relative to the cwd, whereas
+         * the config file basePath can be an ancestor of the cwd.
+         */
+        if (basePath === cwd) {
+            relativeIgnorePatterns = ignorePatterns;
+        } else {
+
+            // relative path must only have Unix-style separators
+            const relativeIgnorePath = path.relative(basePath, cwd).replace(/\\/gu, "/");
+
+            relativeIgnorePatterns = ignorePatterns.map(pattern => {
+                const negated = pattern.startsWith("!");
+                const basePattern = negated ? pattern.slice(1) : pattern;
+
+                return (negated ? "!" : "") +
+                    path.posix.join(relativeIgnorePath, basePattern);
+            });
+        }
```

Comment:
This doesn't work well if `cwd` is not inside `basePath` (which can be the case with the new flag) as the calculated patterns will start with `../` and so will be effectively ignored. But it seems quite complex to fix, so maybe we can leave it for a follow-up PR.

Reply:
Here's the issue: https://github.com/eslint/eslint/issues/18948